File: islamxplorer_flutter/lib/models/hadith.dart
import 'package:islamxplorer_flutter/models/searchResultItem.dart';

class Hadith implements SearchResultItem{
  String id;
  String arabicText;
  String englishText;
  String? narratedBy;
  String source;
  int hadithNo;
  bool isBookmarked = false;
  bool isReported = false;

  @override
  late String sID;

  @override
  late String sTitle;

  @override
  late String sSubtitle;

  @override
  late bool sIsBookmarked;

  @override
  late bool sIsReported;

  @override
  late String sType;

  Hadith({
    required this.id,
    required this.arabicText,
    required this.englishText,
    required this.source,
    this.narratedBy,
    required this.hadithNo
  }) {
    sID = id;
    sTitle = source;
    sSubtitle = englishText;
    sIsBookmarked = isBookmarked;
    sIsReported = isReported;
    sType = runtimeType.toString();
  }

  factory Hadith.fromJson(Map<String, dynamic> json) {
    return Hadith(
      id: json['hadithID'] as String,
      englishText: json['englishText'] as String,
      arabicText: json['arabicText'] as String,
      source: json['source'] as String,
      narratedBy: json['narratedBy'] ?? "",
      hadithNo: json['hadithNo']
    );
  }

  @override
  void updateBookmarkStatus(bool isBookmarked) {
    this.isBookmarked = isBookmarked;
    sIsBookmarked = isBookmarked;
    print("Type: ${sType}");
  }

  @override
  void updateReportStatus(bool isReported) {
    this.isReported = isReported;
    sIsReported = isReported;
    print("Type: ${sType}");
  }
}

File: islamXplorerFlask/models/hadith.py
class Hadith:
    def __init__(self, hadithID, hadithNo=None, englishText=None, embeddings=None, arabicText=None, source=None, narratedBy=None):
        self.id = hadithID
        self.hadithNo = hadithNo
        self.englishText = englishText
        self.arabicText = arabicText
        self.source = source
        self.embeddings = embeddings
        self.narratedBy = narratedBy

    @property
    def __str__(self):
        return "hadithID: " + self.id

    def __repr__(self):
        return self.__str__

    def toEmbeddingsStr(self):
        return {
            "id": self.id,
            "type": "Hadith",
            "embeddings": self.embeddings
        }

    def to_dict(self, embeddings=False):
        hadith_dict = {
            "type": "hadith",
            "hadithID": self.id,
            "arabicText": self.arabicText,
            "englishText": self.englishText,
        }

        if self.source is not None:
            hadith_dict["source"] = self.source
        if self.narratedBy is not None:
            hadith_dict["narratedBy"] = self.narratedBy

        if self.embeddings is not None and embeddings:
            hadith_dict["embeddings"] = self.embeddings

        return hadith_dict
